### PR TITLE
Replace Phobos' parallel function with a hand-written one

### LINKED ISSUE #1181: std::parallel doesn't re-initialize the static variables
std::parallel doesn't re-initialize the static variables, and also doesn't call static destructors, constructors when starting a new task

### PR DESCRIPTION
**Pull request by ferencdg**
fixes #1181

### CODE REVIEW

File: source/agora/test/Base.d [hunk 1]
```diff
@@ -190,11 +192,49 @@ private UnitTestResult customModuleUnitTester ()
     foreach (mod; single_threaded)
         runTest(mod);
 
-    foreach (mod; parallel(parallel_tests))
-        runTest(mod);
+    auto available_cores = new Semaphore(totalCPUs);
```
Comment (AndrejMitrovic) [inline source/agora/test/Base.d:195]: Smart. Love it.

**Review by AndrejMitrovic — APPROVED** [review]
LGTM.

**Review by linked0 — APPROVED** [review]
LGTM

File: source/agora/test/Base.d [hunk 2]
```diff
@@ -190,11 +192,49 @@ private UnitTestResult customModuleUnitTester ()
...
+    auto finished_tasks_num = new Semaphore(0);
+    // we cannot use phobos' parallel function, as that function will not 
+    // re-initialize static variables at the start of a new task
+    void runInParallel (ModTest[] parallel_tests)
+    {
+        class WorkThread : Thread
```
Comment (Geod24) [inline source/agora/test/Base.d:201]: ~static~ Nvm

**Review by Geod24 — APPROVED** [review]
Nice

**Comment by TrustHenry** [comment]
That's great! Can this make Windows parallel testing possible?

**Review by TrustHenry — APPROVED** [review]

File: source/agora/test/Base.d [hunk 3]
```diff
@@ -190,11 +192,49 @@ private UnitTestResult customModuleUnitTester ()
     foreach (mod; single_threaded)
         runTest(mod);
 
-    foreach (mod; parallel(parallel_tests))
-        runTest(mod);
+    auto available_cores = new Semaphore(totalCPUs);
+    auto finished_tasks_num = new Semaphore(0);
+    // we cannot use phobos' parallel function, as that function will not
```
Comment (Geod24) [inline source/agora/test/Base.d]: Trailing whitespaces

File: source/agora/test/Base.d [hunk 4]
```diff
@@ -190,11 +192,49 @@ private UnitTestResult customModuleUnitTester ()
...
+        {
+            auto test = parallel_tests.front;
+            parallel_tests.popFront();
+
+            // wait for a core to become available
+            available_cores.wait();
+
+            (new WorkThread(test)).start();
+        }
+    }
```
Comment (Geod24) [inline source/agora/test/Base.d]: Trailing whitespace

**Review by Geod24 — APPROVED** [review]
Okay the `static` cannot be used here. Removed the trailing whitespaces (please configure your editor), and fixed capitalization of the title. LGTM, will merge when green.

**Comment by ferencdg** [comment]
> That's great! Can this make Windows parallel testing possible?

It depends what the problem with Windows parallel testing was. We no longer call the 'parallel' function, and just depend on the Thread class

**Comment by Geod24** [comment]
I'm 95% certain the answer is no.

**Comment by ferencdg** [comment]
> Okay the `static` cannot be used here. Removed the trailing whitespaces (please configure your editor), and fixed capitalization of the title. LGTM, will merge when green.

done, sorry for that

**Pull request merged**